Connectors/Rhino/Speckle.Connectors.RhinoShared/Bindings/RhinoMapperBinding.cs

```diff
@@ -193,16 +367,168 @@ private void OnDocumentChanged(object? sender, EventArgs e)
+  private static Layer? GetLayer(string layerIdString) =>
+    Guid.TryParse(layerIdString, out var layerId) ? RhinoDoc.ActiveDoc.Layers.FindId(layerId) : null;
+
+  /// <summary>
+  /// Gets the full layer path with / delimiter (matching send workflow).
+  /// Reuses the exact same logic as RhinoLayersFilter.GetFullLayerPath().
+  /// </summary>
+  private static string GetFullLayerPath(Layer layer)
+  {
+    string fullPath = layer.Name;
+    Guid parentIndex = layer.ParentLayerId;
+    while (parentIndex != Guid.Empty)
+    {
+      Layer? parentLayer = RhinoDoc.ActiveDoc.Layers.FindId(parentIndex);
+      if (parentLayer == null)
+      {
+        break;
+      }
+
+      fullPath = parentLayer.Name + "/" + fullPath; // use "/" delimiter like send workflow
+      parentIndex = parentLayer.ParentLayerId;
+    }
+    return fullPath;
+  }
+
+  /// <summary>
+  /// Helper to check if a layer (by index) has a category mapping.
+  /// </summary>
+  private static bool HasLayerMapping(int layerIndex)
+  {
+    if (layerIndex < 0 || layerIndex >= RhinoDoc.ActiveDoc.Layers.Count)
+    {
+      return false;
+    }
+
+    var layer = RhinoDoc.ActiveDoc.Layers[layerIndex];
+    return !string.IsNullOrEmpty(layer.GetUserString(RevitMappingConstants.CATEGORY_USER_STRING_KEY));
+  }
+
+  /// <summary>
+  /// Gets all RhinoObjects in the specified layer and all its child layers recursively.
+  /// </summary>
+  private static IEnumerable<RhinoObject> GetObjectsInLayerHierarchy(Layer rootLayer)
+  {
+    var allObjects = new List<RhinoObject>();
+    var layersToSearch = GetLayerAndAllChildren(rootLayer).ToList();
+
+    foreach (var layer in layersToSearch)
+    {
+      var objectsOnLayer = RhinoDoc.ActiveDoc.Objects.FindByLayer(layer);
+      allObjects.AddRange(objectsOnLayer);
+    }
+
+    return allObjects;
+  }
+
+  /// <summary>
+  /// Gets the specified layer and all its child layers recursively.
+  /// </summary>
+  private static IEnumerable<Layer> GetLayerAndAllChildren(Layer rootLayer)
+  {
+    // Return the root layer itself
+    yield return rootLayer;
+
+    // Get all child layers recursively
+    foreach (var childLayer in GetAllChildLayers(rootLayer))
+    {
+      yield return childLayer;
+    }
+  }
+
+  /// <summary>
+  /// Recursively gets all child layers of the specified parent layer.
+  /// </summary>
+  private static IEnumerable<Layer> GetAllChildLayers(Layer parentLayer)
+  {
+    var doc = RhinoDoc.ActiveDoc;
+    if (doc?.Layers == null)
+    {
+      yield break;
+    }
+
+    // Find all direct child layers
+    var directChildren = doc.Layers.Where(layer => layer.ParentLayerId == parentLayer.Id);
+
+    foreach (var childLayer in directChildren)
+    {
+      // Return the direct child
+      yield return childLayer;
+
+      // Recursively get grandchildren
+      foreach (var grandChild in GetAllChildLayers(childLayer))
+      {
+        yield return grandChild;
+      }
+    }
+  }
+
```

Comment:
looks like all can go into helper class

Reply:
[refactor: consolidate Rhino layer and object utilities into helpers](https://github.com/specklesystems/speckle-sharp-connectors/pull/1027/commits/ee82d2435b27f47e210fc7aae8669e45b369c802)